core/src/main/scala/org/bykn/bosatsu/Evaluation.scala

```diff
@@ -415,39 +458,96 @@ case class Evaluation[T](pm: PackageMap.Typed[T], externals: Externals) {
-                    val itemStr = items.mkString("(", ", ", ")")
-                    sys.error(s"ill typed in match (${ctor.asString}$itemStr\n\n$other\n\nenv: $acc")
-                    // $COVERAGE-ON$
+            if (dt.isNewType) {
+              // we erase this entirely, and just match the underlying item
+              itemFns.head
+            }
+            else if (dt.isStruct) {
+              if (itemsWild) noop
+              else
+                // this is a struct, which means we expect it
+                { (arg: Value, acc: Env) =>
+                  arg match {
+                    case p: ProductValue =>
+                      // this is the pattern
+                      // note passing in a List here we could have union patterns
+                      // if all the pattern bindings were known to be of the same type
+                      processArgs(p.toList, acc)
+
+                    case other =>
+                      // $COVERAGE-OFF$this should be unreachable
+                      val itemStr = items.mkString("(", ", ", ")")
+                      sys.error(s"ill typed in match (${ctor.asString}$itemStr\n\n$other\n\nenv: $acc")
+                      // $COVERAGE-ON$
+                  }
                 }
-              }
             }
             else {
               // compute the index of ctor, so we can compare integers later
               val idx = dt.constructors.map(_._1).indexOf(ctor)
 
-              // we don't check if idx < 0, because if we compiled, it can't be
-              { (arg: Value, acc: Env) =>
-                arg match {
-                  case SumValue(enumId, v) =>
-                    if (enumId == idx) processArgs(v.toList, acc)
-                    else None
-                  case other =>
-                    // $COVERAGE-OFF$this should be unreachable
-                    sys.error(s"ill typed in match: $other")
-                    // $COVERAGE-ON$
-                }
+              getNatLikeFirst(dt) match {
+                case Some(zeroFirst) =>
+                  // we represent Nats with java BigInteger
+                  val isZero = (zeroFirst && (idx == 0)) || (!zeroFirst && (idx == 1))
+
+                  if (isZero)
+                    { (arg: Value, acc: Env) =>
+                      arg match {
+                        case ExternalValue(b: BigInteger) =>
+                          if (b == BigInteger.ZERO) Some(acc)
+                          else None
+                        case other =>
+                          // $COVERAGE-OFF$this should be unreachable
+                          val itemStr = items.mkString("(", ", ", ")")
+                          sys.error(s"ill typed in match (${ctor.asString}$itemStr\n\n$other\n\nenv: $acc")
+                          // $COVERAGE-ON$
+                      }
+                    }
+                  else {
+                    // We are expecting non-zero
+                    val succMatch = itemFns.head
```

Comment:
You use `itemFns.head` twice. Is it possible to get the type system to ensure this is non-empty? Maybe this changes too many things but perhaps a `Pattern.PositionalStruct` with no items should be handled separately?